packages/app/src/background/services/zkIdentity/index.ts

```diff
@@ -273,9 +282,33 @@ export default class ZkIdentityService implements IBackupable {
       });
     }
 
+    this.isUnlocked = true;
+    this.onUnlocked();
+
+    return true;
+  };
+
+  onUnlocked = (): boolean => {
+    if (this.unlockCB) {
+      this.unlockCB();
+      this.unlockCB = undefined;
+    }
+
     return true;
   };
 
+  awaitUnlock = async (): Promise<void> => {
+    if (this.isUnlocked) {
+      return undefined;
+    }
+
+    return new Promise((resolve) => {
+      this.unlockCB = () => {
+        resolve(undefined);
+      };
+    });
+  };
```

Comment:
This code is duplicated. It makes sense to move it to a separate service.

Reply:
Yes, I found that having an abstract class with an abstract impl of lock would make sense than having a composition in this case. 

so that in the injector I can directly call `this.approveService.awaitUnlocked()` not `this.approveService.baseService.awaitUnlocked()`.

Reply:
Yes, in this case abstract class works well.